packages/iov-stream/src/concat.spec.ts

```diff
@@ -0,0 +1,217 @@
+
+    concatenatedStream.addListener({
+      next: value => expect(value).toEqual(expected.shift()!),
+      complete: () => {
+        expect(expected.length).toEqual(0);
+        done();
+      },
+      error: done.fail,
+    });
+  });
+
+  it("changes output order when order of streams switch", done => {
+    const stream1 = Stream.of("1", "2", "3");
+    const stream2 = Stream.of("a", "b", "c");
+    const concatenatedStream = concat(stream2, stream1);
+    const expected = ["a", "b", "c", "1", "2", "3"];
+
+    concatenatedStream.addListener({
+      next: value => expect(value).toEqual(expected.shift()!),
+      complete: () => {
+        expect(expected.length).toEqual(0);
+        done();
+      },
+      error: done.fail,
+    });
+  });
+
+  it("changes output order when order of streams switch", done => {
+    const stream1 = Stream.of("1", "2", "3");
+    const stream2 = Stream.of("a", "b", "c");
+    const concatenatedStream = concat(stream2, stream1);
+    const expected = ["a", "b", "c", "1", "2", "3"];
+
+    concatenatedStream.addListener({
+      next: value => expect(value).toEqual(expected.shift()!),
+      complete: () => {
+        expect(expected.length).toEqual(0);
+        done();
+      },
+      error: done.fail,
+    });
+  });
+
+  it("should concat two asynchronous short streams together", done => {
+    const stream1 = Stream.periodic(25).take(3);
+    const stream2 = Stream.periodic(50).take(2);
+    const concatenatedStream = concat(stream1, stream2);
+    const expected = [0, 1, 2, 0, 1];
+
+    concatenatedStream.addListener({
+      next: value => expect(value).toEqual(expected.shift()!),
+      complete: () => {
+        expect(expected.length).toEqual(0);
+        done();
+      },
+      error: done.fail,
+    });
+  });
+
+  it("should append a synchronous stream after an asynchronous stream", done => {
+    const stream1 = Stream.periodic(25).take(3);
+    const stream2 = Stream.of(30, 40, 50, 60);
+    const concatenatedStream = concat(stream1, stream2);
+    const expected = [0, 1, 2, 30, 40, 50, 60];
+
+    concatenatedStream.addListener({
+      next: value => expect(value).toEqual(expected.shift()!),
+      complete: () => {
+        expect(expected.length).toEqual(0);
+        done();
+      },
+      error: done.fail,
+    });
+  });
+
+  it("buffers asynchonous events of second stream until first stream completes", done => {
+    const sourceStream = Stream.periodic(25);
+    const stream1 = sourceStream.take(3);
+    const stream2 = sourceStream.take(3);
+    const concatenatedStream = concat(stream1, stream2);
+    const expected = [0, 1, 2, 0, 1, 2];
+
+    concatenatedStream.addListener({
+      next: value => expect(value).toEqual(expected.shift()!),
+      complete: () => {
+        expect(expected.length).toEqual(0);
+        done();
+      },
+      error: done.fail,
+    });
+  });
+
+  it("unsubscribes and re-subscribes from source streams", done => {
```

Comment:
Very nice test case...
I just started wondering about that, would producers be registered properly.
Actually, my thought was more like `const combined = concat(stream1, stream2)`, where stream1 is a periodic one (like above), and stream 2 is a producer, that will only produce events when a listener is active.

So...
In the case we listen to stream 1 until done, then listen to stream2, some event will be missing.

Like a bit more complex take on this test.
That events are triggered at certain times and if no listener is registered at time X, then the event gets dropped.

Reply:
What you describe is the test case that differentiates xstream's concat and out concat: https://github.com/staltz/xstream/issues/170#issuecomment-451305237

You can do the same thing with two streams that are both powered by a shared source.

```ts
    const sourceStream = Stream.periodic(25);
    const stream1 = sourceStream.take(3);
    const stream2 = sourceStream.take(3);
    const concatenatedStream = concat(stream1, stream2);
    const expected = [0, 1, 2, 0, 1, 2];
```

If subscription time to `stream2` was not correct, you would get other values than `[0, 1, 2]`.

This is implemented as test "buffers asynchonous events of second stream until first stream completes"

Reply:
Yes. Great example